src/bit-fields-macros/src/builder.rs

```diff
@@ -0,0 +1,880 @@
+    field_matching_from_hashmap: TokenStream,
+    /// String used to define `From<HashMap<String,$data_type>>`.
+    field_setting_hashmap: TokenStream,
+    /// String used to define the table used in the rustdoc for the bit field.
+    struct_doc_table_layout: Vec<TokenStream>,
+    /// Accessor methods to members.
+    struct_accessors: TokenStream,
+    /// String used to form the display the bit field, the lines represent:
+    /// 1. Top border
+    /// 2. Bit numbers
+    /// 3. Border
+    /// 4. Member idents
+    /// 5. Border
+    /// 6. Member values
+    /// 7. Bottom border
+    /// Fmt values (since write doesn't work with in place ones)
+    display_string: MultiLineString,
+    /// String used to pass arguments for `std::fmt::Display` implementation.
+    display_fmt_string: TokenStream,
+    /// Rustdoc to attach to generated bit field.
+    rustdoc: String,
+    /// Struct data type (e.g. `u8`)
+    data_type: DataTypeToken,
+    /// Struct identifier
+    struct_name: Ident,
+    /// Bit flag constants to construct a field with a single flag active.
+    flag_constants: TokenStream,
+    /// Bit mask including all bit flags.
+    bit_flag_mask: TokenStream,
+    /// Bit mask including all bit ranges.
+    bit_range_mask: TokenStream,
+}
+
+impl BitFieldBuilder {
+    /// Constructs new `BitFieldBuilder`.
+    pub fn new(rustdoc: String, struct_name: Ident, data_type: DataTypeToken) -> Self {
+        Self {
+            flag_matching_from_hashset: TokenStream::new(),
+            flag_setting_hashset: TokenStream::new(),
+            field_matching_from_hashmap: TokenStream::new(),
+            field_setting_hashmap: TokenStream::new(),
+            struct_doc_table_layout: vec![quote! {
+                #[doc = "<tr><th>Bit/s</th><th>Identifier</th><th>Description</th></tr>"]
+            }],
+            struct_accessors: TokenStream::new(),
+            #[rustfmt::skip]
+            display_string: MultiLineString::from("\
+                ┌───────\n\
+                │ \x1b[1mBit/s\x1b[0m \n\
+                ├───────\n\
+                │ \x1b[1mDesc\x1b[0m  \n\
+                ├───────\n\
+                │ \x1b[1mValue\x1b[0m \n\
+                └───────",
+            ),
+            display_fmt_string: TokenStream::new(),
+            rustdoc,
+            data_type,
+            struct_name,
+            flag_constants: TokenStream::new(),
+            bit_flag_mask: quote! { 0 },
+            bit_range_mask: quote! { 0 },
+        }
+    }
+
+    /// Adds a bit member to the structure.
+    pub(crate) fn add(self, member: Member) -> Self {
+        match member {
+            Member::BitRange(field) => self.add_bit_range(field),
+            Member::BitFlag(flag) => self.add_bit_flag(flag),
+        }
+    }
+
+    /// Adds a bit range to the structure.
+    pub(crate) fn add_bit_range(
+        mut self,
+        BitRange {
+            range,
+            rustdoc,
+            identifier,
+            skip,
```

Comment:
When converting to/from `HashSet`s and `HashMap`s if fields overlap the value of the resulting bit field may depend on the ordering in which the values in the `HashSet`s and `HashMap`s are iterated over, since this ordering is not consistent this could result in converting from identical `HashSet`s and `HashMap`s producing bit fields of different values. To resolve this, if some defined members of a bit field overlap, `skip` is used on members such that each bit is only contained within 1 non-skipped member.